### ISSUE TITLE
Enforce multiple Authentication Schemes

### ISSUE THREAD

**Issue by oliverjanik** [aspnet/AspNetCore#6733]
When set up with multiple authentication schemes I can do the following to run them both:
```csharp
[Authorize(AuthenticationSchemes = "Scheme1,Scheme2")]
```

This runs both schemes and if at least one succeeds the execution proceeds into the controller action.

I'd like to be able to set this up so that it runs both schemes and makes them both required. I. e. the one that fails issues a challenge.

When both schemes are present the framework correctly merges the ClaimIdentities into User ClaimsPrincipal.

Apologies if this is already possible and I missed it.

**Comment by oliverjanik** [comment]
I found where the schemes get run and principals merged : https://github.com/aspnet/AspNetCore/blob/62351067ff4c1401556725b401478e648b66acdc/src/Security/Authorization/Policy/src/PolicyEvaluator.cs#L33

I've injected my version of PolicyEvaluator to achieve this. This solves my problem.

The question is whether this is something the framework should support with configuration.

**Comment by blowdart** [comment]
I don't think it's enough of a common requirement to make it part of the framework itself

**Comment by onovotny** [comment]
@oliverjanik A blog with a sample that shows what you did would be valuable though.

**Comment by oliverjanik** [comment]
All I did was added early exit from the foreach loop here: https://github.com/aspnet/AspNetCore/blob/62351067ff4c1401556725b401478e648b66acdc/src/Security/Authorization/Policy/src/PolicyEvaluator.cs#L41

I had to, however, copy paste the entire method from original and some internal helper methods too, though.

In my case the schemes were HttpBasic + TLS Client Cert. I can't image this is super rare.

If not out of the box feature, at least the override could be a little easier.

Another idea I had was wrapping them both under custom AuthenticationHandler, which I have not tried yet.
